src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Tooltip/DefaultVSLSPTagHelperTooltipFactory.cs

```diff
@@ -33,87 +45,163 @@ internal class DefaultVSLSPTagHelperTooltipFactory : VSLSPTagHelperTooltipFactor
+        {
+            if (attributeDescriptionInfo is null)
+            {
+                throw new ArgumentNullException(nameof(attributeDescriptionInfo));
+            }
+
+            if (!TryClassifyAttribute(attributeDescriptionInfo, out var descriptionClassifications))
+            {
+                tooltipContent = null;
+                return false;
+            }
+
+            tooltipContent = CombineClassifiedTextRuns(descriptionClassifications, PropertyGlyph);
+            return true;
+        }
+
+        // TO-DO: This method can be removed once LSP's VSCompletionItem supports returning ContainerElements for
+        // its Description property, tracked by https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1319274.
+        public override bool TryCreateTooltip(AggregateBoundElementDescription elementDescriptionInfo, out VSClassifiedTextElement tooltipContent)
+        {
+            if (elementDescriptionInfo is null)
+            {
+                throw new ArgumentNullException(nameof(elementDescriptionInfo));
+            }
+
+            if (!TryClassifyElement(elementDescriptionInfo, out var descriptionClassifications))
+            {
+                tooltipContent = null;
+                return false;
+            }
+
+            tooltipContent = GenerateClassifiedTextElement(descriptionClassifications);
+            return true;
+        }
+
+        // TO-DO: This method can be removed once LSP's VSCompletionItem supports returning ContainerElements for
+        // its Description property, tracked by https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1319274.
+        public override bool TryCreateTooltip(AggregateBoundAttributeDescription attributeDescriptionInfo, out VSClassifiedTextElement tooltipContent)
+        {
+            if (attributeDescriptionInfo is null)
+            {
+                throw new ArgumentNullException(nameof(attributeDescriptionInfo));
+            }
+
+            if (!TryClassifyAttribute(attributeDescriptionInfo, out var descriptionClassifications))
+            {
+                tooltipContent = null;
+                return false;
+            }
+
+            tooltipContent = GenerateClassifiedTextElement(descriptionClassifications);
+            return true;
+        }
+
+        private static bool TryClassifyElement(
+            AggregateBoundElementDescription elementDescriptionInfo,
+            out IReadOnlyList<DescriptionClassification> descriptionClassifications)
+        {
             var associatedTagHelperInfos = elementDescriptionInfo.AssociatedTagHelperDescriptions;
             if (associatedTagHelperInfos.Count == 0)
             {
-                tagHelperDescription = null;
+                descriptionClassifications = null;
                 return false;
             }
 
+            var descriptions = new List<DescriptionClassification>();
```

Comment:
I like this, separate the concern of what data is in there from how it's formatted (the formatting being likely to change later).